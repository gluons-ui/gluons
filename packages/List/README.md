![mandala](https://user-images.githubusercontent.com/1824267/35699021-a5ebb354-0743-11e8-9bf5-9a648a24c7d0.png)
This is a single component of the Mandala UI Kit

Mandala is a React component library that utilizes a functional CSS approach.

https://github.com/mandala-ui/mandala
