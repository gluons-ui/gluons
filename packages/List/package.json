{
  "author": "Lucas K Allmon",
  "description": "Mandala UI List Component",
  "keywords": [
    "react",
    "css",
    "library",
    "components",
    "interface",
    "functional",
    "OOCSS",
    "declarative",
    "ui",
    "user"
  ],
  "license": "MIT",
  "main": "lib/List.js",
  "module": "es/List.js",
  "name": "@mandala-ui/list",
  "publishConfig": {
    "access": "public"
  },
  "scripts": {
    "prepare": "nwb build-react-component"
  },
  "version": "1.0.0"
}
