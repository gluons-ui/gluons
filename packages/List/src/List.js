import React from 'react';
import PropTypes from 'prop-types';
import ListItem from './ListItem';

const List = ({
  children,
  contained,
  indented,
  lineColor,
  lined,
  lineWidth,
  ordered,
  narrow,
  wide,
}) => {
  const childArray = React.Children.toArray(children);
  return (
    <div className="list">
      { ordered ?
        <ol className="list ma0 pa0 tl">
          { childArray.map((child, i, array) => (
            <ListItem
              hasBorder={contained && (i === array.length - 1) ? false : lined}
              indented={indented}
              key={`list-item-${i + 1}`}
              lineColor={lineColor}
              lineWidths={lineWidth}
              narrow={narrow}
              wide={wide}
            >
              {`${i + 1}. `}{child}
            </ListItem>
          ))
          }
        </ol>
        :
        <ul className="list ma0 pa0 tl">
          { childArray.map((child, i, array) => (
            <ListItem
              hasBorder={contained && (i === array.length - 1) ? false : lined}
              indented={indented}
              key={`list-item-${i + 1}`}
              lineColor={lineColor}
              lineWidths={lineWidth}
              narrow={narrow}
              wide={wide}
            >
              {child}
            </ListItem>
          ))
          }
        </ul>
      }
    </div>
  );
};

List.propTypes = {
  children: PropTypes.node,
  contained: PropTypes.bool,
  indented: PropTypes.bool,
  lined: PropTypes.bool,
  lineColor: PropTypes.string,
  lineWidth: PropTypes.number,
  ordered: PropTypes.bool,
  narrow: PropTypes.bool,
  wide: PropTypes.bool,
};

List.defaultProps = {
  children: null,
  contained: false,
  indented: false,
  lined: false,
  lineColor: 'black',
  lineWidth: 0,
  ordered: false,
  narrow: false,
  wide: false,
};

export default List;
