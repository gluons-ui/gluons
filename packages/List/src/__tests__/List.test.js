import React from 'react';
import { shallow } from 'enzyme';
import List from '../List';

describe('<List />', () => {
  it('should exist', () => {
    expect(<List />).toBeDefined();
  });

  it('should exist when called', () => {
    expect(shallow(<List />).exists()).toEqual(true);
  });

  it('should render null when nothing is passed', () => {
    const component = shallow(<List />).childAt(0);
    expect(component.find('list-item').exists()).toEqual(false);
  });

  it('should render a default list', () => {
    const component = shallow(<List />);
    expect(component).toMatchSnapshot();
  });

  it('should render an ordered list', () => {
    const component = shallow(
      <List
        ordered
        lined
        contained
        indented
      >
        <p>A child</p>
        <p>A child</p>
        <p>A child</p>
        <p>A child</p>
        <p>A child</p>
      </List>,
    );
    expect(component).toMatchSnapshot();
  });

  it('should render a group of ListItems when passed children', () => {
    expect(shallow(
      <List>
        <p>A child</p>
        <p>A child</p>
        <p>A child</p>
        <p>A child</p>
        <p>A child</p>
      </List>,
    )).toMatchSnapshot();
  });

  it('should remove the last bottom border if "contained" prop is present', () => {
    expect(shallow(
      <List
        contained
      >
        <p>A child</p>
        <p>A child</p>
        <p>A child</p>
        <p>A child</p>
        <p>A child</p>
      </List>,
    )).toMatchSnapshot();
  });

  it('should add space if the "indented" prop is present', () => {
    expect(shallow(
      <List
        indented
      >
        <p>A child</p>
        <p>A child</p>
        <p>A child</p>
        <p>A child</p>
        <p>A child</p>
      </List>,
    )).toMatchSnapshot();
  });

  it('should render a lined <ListItem /> when prop is present', () => {
    expect(shallow(
      <List
        lined
      >
        <p>A child</p>
        <p>A child</p>
        <p>A child</p>
        <p>A child</p>
        <p>A child</p>
      </List>,
    )).toMatchSnapshot();
  });

  it('should render a custom line color', () => {
    expect(shallow(
      <List
        lineColor="red"
        lined
      >
        <p>A child</p>
        <p>A child</p>
        <p>A child</p>
        <p>A child</p>
        <p>A child</p>
      </List>,
    )).toMatchSnapshot();
  });

  it('should render a custom line width', () => {
    expect(shallow(
      <List
        lineColor="red"
        lined
        lineWidth={2}
      >
        <p>A child</p>
        <p>A child</p>
        <p>A child</p>
        <p>A child</p>
        <p>A child</p>
      </List>,
    )).toMatchSnapshot();
  });
});
