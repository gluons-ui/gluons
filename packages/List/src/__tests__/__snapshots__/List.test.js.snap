// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<List /> should add space if the "indented" prop is present 1`] = `
ShallowWrapper {
  "length": 1,
}
`;

exports[`<List /> should remove the last bottom border if "contained" prop is present 1`] = `
ShallowWrapper {
  "length": 1,
}
`;

exports[`<List /> should render a custom line color 1`] = `
ShallowWrapper {
  "length": 1,
}
`;

exports[`<List /> should render a custom line width 1`] = `
ShallowWrapper {
  "length": 1,
}
`;

exports[`<List /> should render a default list 1`] = `
ShallowWrapper {
  "length": 1,
}
`;

exports[`<List /> should render a group of ListItems when passed children 1`] = `
ShallowWrapper {
  "length": 1,
}
`;

exports[`<List /> should render a lined <ListItem /> when prop is present 1`] = `
ShallowWrapper {
  "length": 1,
}
`;

exports[`<List /> should render an ordered list 1`] = `
ShallowWrapper {
  "length": 1,
}
`;

exports[`<List /> should render less padding with the Narrow prop 1`] = `
ShallowWrapper {
  "length": 1,
}
`;

exports[`<List /> should render more padding with the Wide prop 1`] = `
ShallowWrapper {
  "length": 1,
}
`;
